File: collect_app/src/main/java/org/odk/collect/android/views/ObviousProgressBar.java
package org.odk.collect.android.views;

import android.content.Context;
import android.os.Handler;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;

/**
 * A progress bar that shows for a minimum amount fo time so it's obvious to the user that
 * something has happened.
 */
public class ObviousProgressBar extends ProgressBar {

    public static final int MINIMUM_SHOW_TIME = 750;

    private final Handler handler;

    private Long shownAt;

    public ObviousProgressBar(Context context) {
        super(context);
        handler = new Handler();
    }

    public ObviousProgressBar(@NonNull Context context, AttributeSet attrs) {
        super(context, attrs);
        handler = new Handler();
    }

    public ObviousProgressBar(@NonNull Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        handler = new Handler();
    }

    public void show() {
        handler.removeCallbacksAndMessages(null);
        shownAt = System.currentTimeMillis();
        super.setVisibility(View.VISIBLE);
    }

    public void hide(int visibility) {
        if (shownAt != null) {
            long timeShown = System.currentTimeMillis() - shownAt;

            if (timeShown < MINIMUM_SHOW_TIME) {
                long delay = MINIMUM_SHOW_TIME - timeShown;

                handler.removeCallbacksAndMessages(null);
                handler.postDelayed(() -> makeHiddenOrGone(visibility), delay);
            } else {
                makeHiddenOrGone(visibility);
            }
        } else {
            makeHiddenOrGone(visibility);
        }
    }

    private void makeHiddenOrGone(int visibility) {
        super.setVisibility(visibility);
        shownAt = null;
    }
}

File: androidshared/src/main/java/org/odk/collect/androidshared/ui/ObviousProgressBar.kt
package org.odk.collect.androidshared.ui

import android.content.Context
import android.os.Handler
import android.util.AttributeSet
import com.google.android.material.progressindicator.LinearProgressIndicator

/**
 * A progress bar that shows for a minimum amount fo time so it's obvious to the user that
 * something has happened.
 */
class ObviousProgressBar(
    context: Context,
    attrs: AttributeSet?
) : LinearProgressIndicator(context, attrs) {
    private val handler = Handler()
    private var shownAt: Long? = null

    init {
        super.setVisibility(GONE)
        super.setIndeterminate(true)
    }

    override fun show() {
        handler.removeCallbacksAndMessages(null)
        shownAt = System.currentTimeMillis()
        super.setVisibility(VISIBLE)
    }

    override fun hide() {
        if (shownAt != null) {
            val timeShown = System.currentTimeMillis() - shownAt!!

            if (timeShown < MINIMUM_SHOW_TIME) {
                val delay = MINIMUM_SHOW_TIME - timeShown

                handler.removeCallbacksAndMessages(null)
                handler.postDelayed({ this.makeGone() }, delay)
            } else {
                makeGone()
            }
        } else {
            makeGone()
        }
    }

    private fun makeGone() {
        super.setVisibility(GONE)
        shownAt = null
    }

    companion object {
        private const val MINIMUM_SHOW_TIME = 750
    }
}
